src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+    trace_msg(in, "Variant shift count is not supported");
+    return false;
+  } else {
+    // 5) For other general inputs, include it and also push it into the
+    //    worklist to collect inputs recursively
+    worklist->push(in);
+    stmt->push(in);
+    return true;
+  }
+}
+
+// Collect lists of nodes that make up loop statements
+bool VectorMaskedLoop::collect_statements() {
+  // First, initialize each statement from a store node.
+  for (int idx = 0; idx < _body_nodes.length(); idx++) {
+    Node* node = _body_nodes.at(idx);
+    if (node->is_Store() && supported_mem_access(node->as_Store())) {
+      // Create a new statement and add the store into its node list
+      Node_List* stmt = new Node_List(_arena);
+      stmt->push(node);
+      _stmts.append(stmt);
+    }
+  }
+
+  // Do early bail out if no statement is created
+  int num_stmts = _stmts.length();
+  if (num_stmts == 0) {
+    trace_msg(nullptr, "No vectorizable statement is found");
+    return false;
+  }
+
+  // Then, extend each statement by recursively including input nodes. Bail out
+  // if unsupported node is found.
+  for (int idx = 0; idx < num_stmts; idx++) {
+    Node_List* stmt = _stmts.at(idx);
+    assert(stmt->size() == 1, "Each statement should have exactly one node");
+    Node* store = stmt->at(0);
+    // Add value input of the store node into a worklist to include more nodes
+    // into current statement
+    Node_List* worklist = new Node_List(_arena);
+    if (!collect_statements_helper(store, MemNode::ValueIn, stmt, worklist)) {
+      return false;
+    }
+    // Continue adding nodes until the worklist is empty
+    while (worklist->size() > 0) {
+      Node* node = worklist->pop();
+      uint start, end;
+      VectorNode::vector_operands(node, &start, &end);
+      for (uint idx = start; idx < end; idx++) {
+        if (!collect_statements_helper(node, idx, stmt, worklist)) {
+          return false;
+        }
+      }
+    }
+#ifndef PRODUCT
+    if (TraceMaskedLoop) {
+      tty->print_cr("Nodes in statement [%d] with element type '%s'",
+                    idx, type2name(statement_bottom_type(stmt)));
+      for (uint i = 0; i < stmt->size(); i++) {
+        stmt->at(i)->dump();
+      }
+      tty->cr();
+    }
+#endif
+  }
+
+  return true;
+}
+
+// Analyze loop statements and bail out if any of them is not vectorizable
+bool VectorMaskedLoop::analyze_vectorizability() {
+  if (!find_vector_element_types()) {
+    return false;
+  }
+  if (!vector_nodes_implemented()) {
+    return false;
+  }
+  // Delegate data dependence check to SWPointer utility
+  if (SWPointer::has_potential_dependence(_swptrs)) {
+    trace_msg(nullptr, "Potential data dependence is found in the loop");
+    return false;
+  }
+  if (!analyze_loop_body_nodes()) {
+    return false;
+  }
+  return true;
+}
+
+// Find element basic type for each vectorization candidate node
+bool VectorMaskedLoop::find_vector_element_types() {
```

Comment:
This is very similar to `SuperWord::compute_vector_element_type`. It would be nice to extract it from both and have some shared utility, right?

Reply:
Or is there a clear reason why the two are too different?

Reply:
We need more investigation and discussions about this. Will discuss with you later.

Reply:
Sounds good.